karmada operator: add karmada operator readme (#3507)
karmada-io/karmada: operator/README.md

karmada-demo-etcd-0                                     1/1     Running   0          37s
karmada-demo-kube-controller-manager-584f978bbd-fftwq   1/1     Running   0          5s
karmada-demo-scheduler-6d77b7547-hgz8n                  1/1     Running   0          5s
karmada-demo-webhook-6f5944f5d8-bpkqz                   1/1     Running   0          5s
```

## Upgrade a Karmada instance

Once a Karmada instance is created, the CR resource is automatically filled with default values.
To upgrade the Karmada instance, for example, you can upgrade the Karmada version to v1.5.0 or newer:

```shell
kubectl patch karmada karmada-demo -n test --type merge -p '
{
  "spec": {
    "components": {
      "karmadaAggregatedAPIServer": {
        "imageTag": "v1.5.0"
      },
      "karmadaControllerManager": {
        "imageTag": "v1.5.0"
      },
      "karmadaScheduler": {
        "imageTag": "v1.5.0"
      },
      "karmadaWebhook": {
        "imageTag": "v1.5.0"
      }
    }
  }
}'
```

## Delete a Karmada instance

A Karmada instance can be deleted once the Karmada CR is deleted.

```shell
kubectl delete karmada karmada-demo -n test
```

## Custom Karmada CR

> This feature allows you to configure the Karmada CR to install Karmada instances flexibly. For details, see [karmada.yaml](./config/samples/karmada.yaml).

Reviewer: Could we add some more use cases?
eg. install karmada-descheduler/custom labels.

# Karmada operator

## Overview

> The Karmada operator is a project of the Karmada community designed to manage the lifecycle of Karmada. we can operate CR to install, upgrade and delete the Karmada instance. It is a centrally managed controller that can install karmada instance not only in the local cluster, but also in a remote cluster. It is convenient to manage multiple karmada instances in a global cluster.

Reviewer: We can

> The Karmada operator is a project of the Karmada community designed to manage the lifecycle of Karmada. we can operate CR to install, upgrade and delete the Karmada instance. It is a centrally managed controller that can install karmada instance not only in the local cluster, but also in a remote cluster. It is convenient to manage multiple karmada instances in a global cluster.

Reviewer: We need to unify the `Karmada instance` and `karmada instance`.

Author: I'm with you.

## Getting started

### Quik start

Create a secret of kubeconfig to connect cluster.

```shell
kubectl create namespace karmada-system
kubectl create secret generic my-kubeconfig --from-file=$HOME/.kube/config -n karmada-system
```

Deploy karmada operator workload.

```shell
kubectl apply -f operator/config/deploy/karmada-operator.yaml
```

The pod of Karmada operator in `karmada-system` namespace will be running.

```shell
kubectl get po -n karmada-system
NAME                               READY   STATUS    RESTARTS   AGE
karmada-operator-5b7f485c5-g5lj5   1/1     Running   0          26s
```

## Create Karmada

> `Karmada` CR represents a karmada instance, which is a Namespace-level resource. <br>

Reviewer: We need to unify `Karmada` CR or Karmada CR or karmada CR.

karmada-demo-controller-manager-64455f7fd4-stls6        1/1     Running   0          5s
karmada-demo-etcd-0                                     1/1     Running   0          37s
karmada-demo-kube-controller-manager-584f978bbd-fftwq   1/1     Running   0          5s
karmada-demo-scheduler-6d77b7547-hgz8n                  1/1     Running   0          5s
karmada-demo-webhook-6f5944f5d8-bpkqz                   1/1     Running   0          5s
```

## Upgrade Karmada

After the Karmada CR is created, the CR resource is automatically filled with defaults. <br>
Now update Karmada CR to upgrade karmada version to v1.5.0:

```shell
kubectl patch karmada karmada-demo -n test --type merge -p '
{
  "spec": {
    "components": {
      "karmadaAggregatedAPIServer": {
        "imageTag": "v1.5.0"
      },
      "karmadaControllerManager": {
        "imageTag": "v1.5.0"
      },
      "karmadaScheduler": {
        "imageTag": "v1.5.0"
      },
      "karmadaWebhook": {
        "imageTag": "v1.5.0"
      }
    }
  }
}'
```

## Uninstall Karmada

Delete Karmada CR, the karmada instance will be deleted together.

```shell
kubectl delete karmada karmada-demo -n test
```

> ## Custom Karmada CR
> 
> Flexibility to configure CR to install Karmada instances. For details, please See [Karmada.yaml](./config/samples/karmada.yaml) <br>
> 
> ### set karmada component replica
> 
> The replica of all karmada components is modifiable. scale etcd pod replica to 3, e.g:

Reviewer 2: ```suggestion
## Custom Karmada CR

This feature allows you to configure the Karmada CR to install Karmada instances flexibly. For details, see [karmada.yaml](./config/samples/karmada.yaml).

### Set Karmada component replicas

The replicas of all Karmada components can be modified. For example, to scale the etcd pod replicas to 3:
```

```yaml
apiVersion: operator.karmada.io/v1alpha1
kind: Karmada
metadata:
  name: karmada-demo
  namespace: test
spec:
  components:
    etcd:
      local:
        replicas: 3
```

> ### karmada-apiserver service type

Reviewer 2: ```suggestion
### Change karmada-apiserver service type
```

> The service type of karmada-apiserver is `ClusterIP` by default, update it to `NodePort`, e.g:

Reviewer 2: ```suggestion
The service type of karmada-apiserver is `ClusterIP` by default. You can change it to `NodePort`:
```

```yaml
...
karmadaAPIServer:
  imageRepository: registry.k8s.io/kube-apiserver
  imageTag: v1.25.4
  replicas: 1
  serviceType: NodePort
  serviceSubnet: 10.96.0.0/12
...
```

> ### karmada-apiserver SANs

Reviewer 2: ```suggestion
### Add karmada-apiserver SANs
```

> Add extra SANs to karmada-apiserver certificate. e.g:

Reviewer 2: ```suggestion
You can add more SANs to karmada-apiserver certificate:
```
